2001-02-16  Jakub Jelinek  <jakub@redhat.com>

	* call.c (convert_like_real): Create a temporary for non-lvalue.

	* g++.old-deja/g++.other/init16.C: Update the test so that it does
	not need <string> and also tests the initialization at runtime.

2000-11-08  Jakub Jelinek  <jakub@redhat.com>

	* g++.old-deja/g++.other/init16.C: New test.

--- gcc/cp/call.c.jj	Fri Feb  9 01:35:16 2001
+++ gcc/cp/call.c	Sat Feb 17 00:38:33 2001
@@ -3772,7 +3772,7 @@ convert_like_real (convs, expr, fn, argn
 	tree ref_type = TREE_TYPE (convs);
 
 	/* If necessary, create a temporary.  */
-	if (NEED_TEMPORARY_P (convs))
+	if (NEED_TEMPORARY_P (convs) || !lvalue_p (expr))
 	  {
 	    tree type = TREE_TYPE (TREE_OPERAND (convs, 0));
 	    expr = build_target_expr_with_type (expr, type);
--- gcc/testsuite/g++.old-deja/g++.other/init16.C.jj	Sat Feb 17 00:40:33 2001
+++ gcc/testsuite/g++.old-deja/g++.other/init16.C	Sat Feb 17 00:40:27 2001
@@ -0,0 +1,28 @@
+// Origin: Jakub Jelinek <jakub@redhat.com>
+
+struct bar {
+  char c;
+  bar (const char *);
+  bar (const bar &);
+};
+
+struct foo {
+  bar x;
+};
+
+extern const struct foo y = { "foo" };
+
+bar::bar (const bar &ref)
+{
+  c = ref.c;
+}
+
+bar::bar (const char *p)
+{
+  c = p[2];
+}
+
+int main ()
+{
+  return y.x.c != 'o';
+}
